taiga_halo2/src/note.rs

```diff
@@ -0,0 +1,150 @@
+use crate::{
+    constant::{BASE_BITS_NUM, NOTE_COMMITMENT_R_GENERATOR, NOTE_COMMIT_DOMAIN},
+    nullifier::Nullifier,
+    token::Token,
+    user::User,
+    utils::{extract_p, poseidon_hash},
+};
+use bitvec::{array::BitArray, order::Lsb0};
+use core::iter;
+use ff::{Field, PrimeFieldBits};
+use group::{cofactor::CofactorCurveAffine, Group};
+use pasta_curves::pallas;
+use rand::{Rng, RngCore};
+
+/// A commitment to a note.
+#[derive(Copy, Debug, Clone)]
+pub struct NoteCommitment(pallas::Point);
+
+impl NoteCommitment {
+    pub fn inner(&self) -> pallas::Point {
+        self.0
+    }
+
+    pub fn get_x(&self) -> pallas::Base {
+        extract_p(&self.0)
+    }
+}
+
+impl Default for NoteCommitment {
+    fn default() -> NoteCommitment {
+        NoteCommitment(pallas::Point::generator())
+    }
+}
+
+/// A note
+#[derive(Debug, Clone, Default)]
+pub struct Note {
+    /// Owner of the note
+    pub user: User,
+    pub token: Token,
+    pub value: u64,
+    /// for NFT or whatever. TODO: to be decided the value format.
+    pub data: pallas::Base,
+    /// old nullifier. Nonce which is a deterministically computed, unique nonce
+    pub rho: Nullifier,
+    /// computed from spent_note_nf and rcm by using a PRF
+    pub psi: pallas::Base,
+    pub rcm: pallas::Scalar,
+}
+
+impl Note {
+    pub fn new(
+        user: User,
+        token: Token,
+        value: u64,
+        rho: Nullifier,
+        data: pallas::Base,
+        rcm: pallas::Scalar,
+    ) -> Self {
+        let psi = Self::derive_psi(&rho.inner(), &rcm);
+        Self {
+            user,
+            token,
+            value,
+            data,
+            rho,
+            psi,
+            rcm,
+        }
+    }
+
+    // psi = poseidon_hash(rho, (rcm * generator).x)
+    // The psi derivation is different from Orchard, in which psi = blake2b(rho||rcm)
```

Comment:
They even didn't constrain the `psi = blake2b(rho||rcm)` in circuit and used the `psi` as a witness input directly.

Reply:
I think if the prover chooses psi adversarially (against the protocol), it shouldn't hurt anyone else.
But we should document and review this assumption because the prover might not be the note owner, in case we do need to check psi in the circuit.

Reply:
In Orchard `cm = SinsemillaCommit(user_address || app_address || data || rho || psi || value, rcm)`, `psi = prf([9] || rho || rseed)` and `rcm = prf([5] || rho || rseed)`, I didn't figure out why we need the `psi`. Can we get rid of it?